### PR TITLE
Client config functional options

### PR DESCRIPTION
**Pull request by jdziurzynski**
KNTK - 383

Implemented functional options for client config structure.

Design doc: https://docs.google.com/document/d/1b-huL-uHSVyI3eaao3-YOFHOZ115Sz9FOjlNOdHIULY/edit#heading=h.y6dsqy4lm0p4

### CODE REVIEW

File: kentikapi/cloud_exports_integration_test.go
Comment (mateuszmidor) [inline kentikapi/cloud_exports_integration_test.go]: not sure here, shouldn't it be always both email+token or nothing? Like user+password?

Comment (jdziurzynski) [reply kentikapi/cloud_exports_integration_test.go]: I think Your right. If Daniel have nothing against I will merge those two options.

File: kentikapi/client.go
Comment (danielfurman) [inline kentikapi/client.go]: I think that it would be better to split WithRetryConfig to multiple simple options, because RetryConfig contains pointer fields, which are not easy to use.
E.g.:
- WithRetryMaxAttempts(maxAttempts uint)
- WtihRetryMinDelay(minDelay time.Duration)
- WithRetryMaxDelay(maxDelay time.Duration)

What do you think?

**Review by mmac-m3a — APPROVED** [review]
LGTM. I'll wait for a review from @mateuszmidor merging.

File: kentikapi/client.go
Comment (danielfurman) [inline kentikapi/client.go]: I suggest to replace WithAuthEmail() and WithAuthToken with WithCredentials()

File: kentikapi/client.go
Comment (danielfurman) [inline kentikapi/client.go]: This struct stopped being part of public API, so should be un-exported. Its fields might stay exported.

File: kentikapi/client.go
Comment (danielfurman) [inline kentikapi/client.go]: Those options are part of public interface and need doc-strings. Please move doc-strings from config struct fields. Please include default values too.

File: kentikapi/cloud_exports_integration_test.go
Comment (danielfurman) [reply kentikapi/cloud_exports_integration_test.go]: No point in waiting for me. Just do it ;)

File: kentikapi/client.go
Comment (danielfurman) [inline kentikapi/client.go]: I see that this branch includes suggestions from previous PR and I see larger diff than optimal . Please rebase the branch onto dev_grpc.
```
 * f75b0ab  (upstream/dev_grpc) TestClient_GetAgentWithRetries erro timeout fix (#92) (10 hours ago) <Jan Dziurzyński>
* 2e6fa67  Modify cloud export properties field to emulate union (#89) (14 hours ago) <Daniel Furman>
* 7854f6b  Fix payloads apiv6 (#85) (14 hours ago) <Jan Dziurzyński>
* e13b3c5  Implemeted suggestions from review (#86) (14 hours ago) <Jan Dziurzyński>
| * 66542ed  (HEAD -> client_functional_options) Client config functional options (5 days ago) <Jan Dziurzynski>
| * ae71411  Implemented suggetsios from review (5 days ago) <Jan Dziurzynski>
|/  
* 44fc3a9  (dev_grpc) RequiredFields structs for constructors of models (#82) (7 days ago) <Jan Dziurzyński>
```

File: kentikapi/internal/testutil/http.go
Comment (danielfurman) [inline kentikapi/internal/testutil/http.go]: Those are old PR pre-review changes and they shouldn't be included.

File: kentikapi/client.go
Comment (danielfurman) [inline kentikapi/client.go]:
```suggestion
func WithTimeout(timeout time.Duration) ConfigOption {
```
It doesn't make sense to take a pointer here

File: kentikapi/client.go
Comment (danielfurman) [inline kentikapi/client.go]: Please just initialize it here with default values (or call config constructor). No need to split it to two steps now.

File: kentikapi/client.go
Comment (danielfurman) [inline kentikapi/client.go]: Pointer is unnecessary now.

**Review by danielfurman — CHANGES_REQUESTED** [review]

File: kentikapi/client.go
Comment (danielfurman) [inline kentikapi/client.go]:
```suggestion
type ClientOption func(*Config)
```
It's an option to the client.

File: kentikapi/client.go
Comment (jdziurzynski) [reply kentikapi/client.go]: Ok, it seems convenient for the user to use.

**Comment by danielfurman** [comment]
@jdziurzynski FYI: there is no need to name commit "[Rebase & implemented suggestions from review](https://github.com/kentik/community_sdk_golang/pull/91/commits/8527ef4094f27041e1f52d10b5956137c420666c)" I mean "rebase" in commit name does not make sense, becuase it is a new commit. Previous commit was indeed rebased, but we can see that, becuase it has new hash.

File: kentikapi/client.go
Comment (danielfurman) [inline kentikapi/client.go]:
```suggestion
// WithAPIURL sets the Kentik API URL. Default: "https://api.kentik.com".
```

File: kentikapi/client.go
Comment (danielfurman) [inline kentikapi/client.go]:
```suggestion
		c.Timeout = &timeout
```
This is a bug, because that way user cannot set 0 timeout (no timeout).

File: kentikapi/client.go
Comment (danielfurman) [inline kentikapi/client.go]:
```suggestion
// WithCredentials specifies authentication email and token for the Client.
```

File: kentikapi/client.go
Comment (danielfurman) [inline kentikapi/client.go]:
```suggestion
		c.RetryCfg.MaxAttempts = &maxAttempts
```

This is a bug, because that way user cannot set 0 maxAttempts (infinite).

File: kentikapi/client.go
Comment (danielfurman) [inline kentikapi/client.go]:
```suggestion
		c.RetryCfg.MinDelay = &minDelay
```
This is a bug, because that way user cannot set 0 minDelay.

File: kentikapi/client.go
Comment (danielfurman) [inline kentikapi/client.go]:
```suggestion
		c.RetryCfg.MaxDelay = &maxDelay
```
This is a bug, because that way user cannot set 0 maxDelay.

**Review by danielfurman — CHANGES_REQUESTED** [review]

File: kentikapi/client.go
Comment (danielfurman) [reply kentikapi/client.go]: Done

File: kentikapi/client.go
Comment (danielfurman) [reply kentikapi/client.go]: Done

File: kentikapi/cloud_exports_integration_test.go
Comment (danielfurman) [reply kentikapi/cloud_exports_integration_test.go]: Done

File: kentikapi/client.go
Comment (danielfurman) [reply kentikapi/client.go]: Done

File: kentikapi/client.go
Comment (danielfurman) [reply kentikapi/client.go]: Done

File: kentikapi/client.go
Comment (danielfurman) [reply kentikapi/client.go]: Done

File: kentikapi/internal/testutil/http.go
Comment (danielfurman) [reply kentikapi/internal/testutil/http.go]: Done

File: kentikapi/client.go
Comment (danielfurman) [reply kentikapi/client.go]: Done

File: kentikapi/client.go
Comment (danielfurman) [reply kentikapi/client.go]: Done

File: kentikapi/client.go
Comment (danielfurman) [reply kentikapi/client.go]: Done

File: kentikapi/client.go
Comment (danielfurman) [reply kentikapi/client.go]: Done

File: examples/utils.go
Comment (danielfurman) [inline examples/utils.go]: I think that examples should have log payloads disabled, because they print needed outputs.
```suggestion
```

File: examples/utils.go
Comment (danielfurman) [inline examples/utils.go]:
```suggestion
	return kentikapi.NewClient(
```

File: kentikapi/client.go
Comment (danielfurman) [inline kentikapi/client.go]:
```suggestion
```

File: kentikapi/client.go
Comment (danielfurman) [inline kentikapi/client.go]:
```suggestion
```

File: kentikapi/client_integration_test.go
Comment (danielfurman) [inline kentikapi/client_integration_test.go]: This subtest is broken now - it does not verify any retry. Please remember that the purpose of tests is to verify particular use-case of the client. Not to pass. Clearly `A Timeout of zero means no timeout.` docstring from WithTimeout() func is not true and should be removed.
Please use default timeout in subtests, that do not need to modify the timeout. It can be done similarly to options in TestClient_GetAgentWithRetries

File: kentikapi/client_integration_test.go
Comment (danielfurman) [inline kentikapi/client_integration_test.go]: Please use default timeout here too. Using 5 Millisecond might be flaky and is confusing

File: kentikapi/client_integration_test.go
Comment (danielfurman) [inline kentikapi/client_integration_test.go]: This test tests that default timeout is sane. Please do not override it, because it destroys the purpose of the test.

**Review by danielfurman — CHANGES_REQUESTED** [review]

File: kentikapi/client.go
Comment (danielfurman) [reply kentikapi/client.go]: Done

File: kentikapi/client.go
Comment (danielfurman) [reply kentikapi/client.go]: Done. However underlying code does not handle 0 as no timeout so I added suggestion to remove that docstring.

File: kentikapi/client.go
Comment (danielfurman) [reply kentikapi/client.go]: Done

File: kentikapi/client.go
Comment (danielfurman) [reply kentikapi/client.go]: Done

File: kentikapi/client.go
Comment (danielfurman) [reply kentikapi/client.go]: Done

File: kentikapi/client.go
Comment (danielfurman) [reply kentikapi/client.go]: Done

File: examples/utils.go
Comment (danielfurman) [reply examples/utils.go]: Done

File: examples/utils.go
Comment (danielfurman) [reply examples/utils.go]: Done

File: kentikapi/client.go
Comment (danielfurman) [reply kentikapi/client.go]: Done

File: kentikapi/client.go
Comment (danielfurman) [reply kentikapi/client.go]: Done

File: kentikapi/client_integration_test.go
Comment (danielfurman) [reply kentikapi/client_integration_test.go]: Done

File: kentikapi/client_integration_test.go
Comment (danielfurman) [reply kentikapi/client_integration_test.go]: Done

File: kentikapi/client_integration_test.go
Comment (danielfurman) [reply kentikapi/client_integration_test.go]: Done

**Review by danielfurman — APPROVED** [review]

**Pull request merged**
